File: misc/aooswd/ex/command/command.cpp
class Command
{
public:
	virtual ~Command ();
	virtual void Execute () = 0;
protected:
	Command ();
};

class OpenCommand: public Command
{
public:
	OpenCommand (Application*);
	virtual void Execute ();
protected:
	virtual const char* AskUser ();
private:
	Application* _application;
	char* _response;
};

OpenCommand::OpenCommand (Appliation* a)
{
	_application = a;
}

void OpenCommand::Execute ()
{
	const char* name = AskUser ();
	if (name != 0)
	{
		Document* document = new Document (name);
		_application -> Add (document);
		document -> Open ();
	}
}

class PasteCommand: public Command
{
public:
	PasteCommand (Document*);
	virtual void Execute ();
private:
	Document* _document;
};

PasteCommand::PasteCommand (Document* doc)
{
	_document = doc;
}

void PasteCommand::Execute ()
{
	_document -> Paste ();
}

template <class Receiver>
class SimpleCommand: public Command
{
public:
	typedef void (Receiver::*Action) ();
	SimpleCommand (Receiver* r, Action a): _receiver (r), _action (a) {}
	virtual void Execute ();
private:
	Action _action;
	Receiver* _receiver;
};

template <class Receiver>
void SimpleCommand<Receiver>::Execute ()
{
	(_receiver -> *_action)();
}

MyClass* receiver = new MyClass;
Command* aCommand = new SimpleCommand<MyClass>(receiver, &MyClass::Action);
aCommand -> Execute ();

class MacroCommand: public Command
{
public:
	MacroCommand ();
	virtual ~MacroCommand ();
	virtual void Add (Command*);
	virtual void Remove (Command*);
	virtual void Execute ();
private:
	List<Command*>* _cmds;
};

void MacroCommand::Execute ()
{
	ListIterator<Command*> I(_cmds);
	for (I.First (); !I.IsDone (); I.Next ())
	{
		Command* c = i.CurrentItem ();
		c -> Execute ();
	}
}

void MacroCommand::Add (Command* c)
{
	_cmds -> Append ();
}

void MacroCommand::Remove (Command* c)
{
	_cmds -> Remove (c);
}

File: misc/aooswd/ex/command/command.java
public abstract class Command
{
	protected Command ()
	{
	}

	public abstract void Execute();
}

public class OpenCommand extends Command
{
	private Application _application;

	private String _response;

	public OpenCommand (Application A)
	{
		_application = A;
	}

	public void Execute()
	{
		String name = AskUser ();
		if (name != "")
		{
			Document document = new Document (name);
			_application.Add (document);
			document.Open();
		}
	}

	protected String AskUser();

}

public class PasteCommand extends Command
{
	private Document _document;

	public PasteCommand (Document D)
	{
		_document = D;
	}


	public void Execute()
	{
		_document.Paste();
	}

}

public class SimpleCommand extends Command
{
	private Action _action;
	private Object _receiver;

	public SimpleCommand (Object r, Action a)
	{
		_receiver = r;
		_action = a;
	}

	public void execute ()
	{
		_receiver._action();
	}
}

public class MacroCommand extends Command
{
	private LinkedList _cmds;

	public MacroCommand ()
	{
		_cmds = new LinkedList();
	}

	public void Add (Command C)
	{
		_cmds.addLast (C);
	}

	public void Remove (Command C)
	{
		_cmds.remove(C);
	}

	public void execute ()
	{
		ListIterator I = _cmds.listIterator();
		while (_cmd.hasNext())
			((Command)(_cmd.next())).execute();
	}
}

MyClass receiver = new MyClass();
Command aCommand = new SimpleCommand (receiver, Action);
aCommand.execute();




